backend/repository/client_update.go

```diff
@@ -0,0 +1,179 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package repository
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+
+	"github.com/pkg/errors"
+)
+
+// ClientJSONConfig defines the JSON content of the configuration file.
+type ClientJSONConfig struct {
+	Name   string       `json:"name"`
+	Type   string       `json:"type"`
+	Config *configField `json:"config"`
+}
+
+type configField struct {
+	ThriftFile     string            `json:"thriftFile"`
+	ThriftFileSha  string            `json:"thriftFileSha"`
+	ExposedMethods map[string]string `json:"exposedMethods,omitempty"`
+}
+
+// NewClientConfigJSON converts ClientConfig to ClientJSONConfig.
+func NewClientConfigJSON(cfg ClientConfig) *ClientJSONConfig {
+	cfgJSON := &ClientJSONConfig{
+		Name: cfg.Name,
+		Type: string(cfg.Type),
+		Config: &configField{
+			ThriftFile:     cfg.ThriftFile,
+			ExposedMethods: cfg.ExposedMethods,
+		},
+	}
+	return cfgJSON
+}
+
+// UpdateClientConfigs updates JSON configuration files for a client update request.
+func (r *Repository) UpdateClientConfigs(req *ClientConfig, clientCfgDir, thriftFileSha string) error {
+	if err := validateClientUpdateRequest(req); err != nil {
+		return err
+	}
+	cfgJSON := NewClientConfigJSON(*req)
+	cfgJSON.Config.ThriftFileSha = thriftFileSha
+	clientPath := filepath.Join(r.absPath(clientCfgDir), cfgJSON.Name)
+	r.Lock()
+	defer r.Unlock()
+	if err := os.MkdirAll(clientPath, os.ModePerm); err != nil {
+		return errors.Wrap(err, "failed to create client config dir")
+	}
+	if err := writeToJSONFile(filepath.Join(clientPath, clientConfigFileName), cfgJSON); err != nil {
+		return errors.Wrapf(err, "failed to write config for the client %q", cfgJSON.Name)
+	}
+	if err := WriteClientModuleJSON(r.absPath(clientCfgDir)); err != nil {
+		return errors.Wrap(err, "failed to write module config for all clients")
+	}
+	if err := UpdateProductionConfigJSON(req, r.absPath(productionCfgJSONPath)); err != nil {
+		return errors.Wrap(err, "failed to update gateway production config")
+	}
+	return nil
+}
+
+func validateClientUpdateRequest(req *ClientConfig) error {
+	if len(req.ExposedMethods) == 0 {
+		return errors.New("invalid request: no method is exposed for the client")
+	}
+	if req.Type == "tchannel" && req.MuttleyName == "" {
+		return errors.New("invalid request: muttley name is required for tchannel client")
+	}
+	if req.IP == "" || req.Port == 0 {
+		return errors.New("invalid request: ip and port are required")
+	}
+	if req.Type == "tchannel" && req.Timeout == 0 {
+		req.Timeout = 10000
+	}
+	if req.Type == "tchannel" && req.TimeoutPerAttempt == 0 {
+		req.TimeoutPerAttempt = 10000
+	}
+	return nil
+}
+
+type clientModuleJSONConfig struct {
```

Comment:
Dependencies are took care of at module level, we can add it to the clientConfig struct